cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -663,23 +663,23 @@ private ApplicationId updateApplicationByArtifact(ApplicationId appId,
                                                                                    newArtifactId));
     ArtifactDetail newArtifactDetail = artifactRepository.getArtifact(newArtifact);
 
-    return updateApplicationInternal(appId, appSpec.getConfiguration(), programId -> { }, newArtifactDetail,
+    return updateApplicationInternal(appId, programId -> { }, newArtifactDetail,
                                      Collections.singletonList(ApplicationConfigUpdateAction.UPGRADE_ARTIFACT),
-                                     allowedArtifactScopes, allowSnapshot, ownerAdmin.getOwner(appId));
+                                     allowedArtifactScopes, allowSnapshot, ownerAdmin.getOwner(appId), appSpec);
   }
 
   /**
    * Updates an application config by applying given update actions. The app should know how to apply these actions
    * to its config.
    */
   private ApplicationId updateApplicationInternal(ApplicationId appId,
-                                                  @Nullable String currentConfigStr,
                                                   ProgramTerminator programTerminator,
                                                   ArtifactDetail artifactDetail,
                                                   List<ApplicationConfigUpdateAction> updateActions,
                                                   Set<ArtifactScope> allowedArtifactScopes,
                                                   boolean allowSnapshot,
-                                                  @Nullable KerberosPrincipalId ownerPrincipal) throws Exception {
+                                                  @Nullable KerberosPrincipalId ownerPrincipal,
+                                                  @Nullable ApplicationSpecification appSpec) throws Exception {
```

Comment:
please remove `@Nullable` from appSpec as we are calling `appSpec.getConfiguration()` without checking for nullability (even in the previous code)